agent/internal/taskoutput/test_log.go

```diff
@@ -0,0 +1,299 @@
+				Priority:  level.Info,
+				Timestamp: time.Now().UnixNano(),
+				Data:      line,
+			})
+		}
+	}
+
+	return tsk.TaskOutputInfo.TestLogs.Append(ctx, taskoutput.TaskOptions{
+		ProjectID: tsk.Project,
+		TaskID:    tsk.Id,
+		Execution: tsk.Execution,
+	}, testLog.Name, lines)
+}
+
+// TODO (DEVPROD-75): Remove this logic once we cut over to Evergreen logs.
+func sendTestLogToCedar(ctx context.Context, comm client.Communicator, tsk *task.Task, testLog *testlog.TestLog) error {
+	conn, err := comm.GetCedarGRPCConn(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "getting the Cedar gRPC connection for test '%s'", testLog.Name)
+	}
+
+	timberOpts := &buildlogger.LoggerOptions{
+		Project:    tsk.Project,
+		Version:    tsk.Version,
+		Variant:    tsk.BuildVariant,
+		TaskName:   tsk.DisplayName,
+		TaskID:     tsk.Id,
+		Execution:  int32(tsk.Execution),
+		TestName:   testLog.Name,
+		Mainline:   !tsk.IsPatchRequest(),
+		Storage:    buildlogger.LogStorageS3,
+		ClientConn: conn,
+	}
+	levelInfo := send.LevelInfo{Default: level.Info, Threshold: level.Debug}
+	sender, err := buildlogger.NewLoggerWithContext(ctx, testLog.Name, levelInfo, timberOpts)
+	if err != nil {
+		return errors.Wrapf(err, "creating buildlogger logger for test result '%s'", testLog.Name)
+	}
+
+	sender.Send(message.ConvertToComposer(level.Info, strings.Join(testLog.Lines, "\n")))
+	if err = sender.Close(); err != nil {
+		return errors.Wrapf(err, "closing buildlogger logger for test result '%s'", testLog.Name)
+	}
+
+	return nil
+}
+
+type testLogDirectoryHandler struct {
+	dir             string
+	logger          client.LoggerProducer
+	watcher         *watcher.Watcher
+	spec            testLogSpec
+	createSender    func(context.Context, string) (send.Sender, error)
+	followFileCount int
+
+	once sync.Once
+	wg   sync.WaitGroup
+}
+
+func newTestLogDirectoryHandler(output taskoutput.TestLogOutput, taskOpts taskoutput.TaskOptions, logger client.LoggerProducer) *testLogDirectoryHandler {
+	h := &testLogDirectoryHandler{logger: logger}
+	h.createSender = func(ctx context.Context, logPath string) (send.Sender, error) {
+		return output.NewSender(ctx, taskOpts, taskoutput.EvergreenSenderOptions{
+			Local: logger.Task().GetSender(),
+			Parse: h.spec.getParser(),
+		}, logPath)
+	}
+
+	return h
+}
+
+func (h *testLogDirectoryHandler) start(ctx context.Context, dir string) error {
+	h.dir = dir
+	h.watcher = watcher.New()
+	h.watcher.FilterOps(watcher.Create)
+	if err := h.watcher.AddRecursive(h.dir); err != nil {
+		return errors.Wrap(err, "configuring test log directory watcher as recursive")
+	}
+	if err := h.watcher.Ignore(filepath.Join(h.dir, testLogSpecFilename)); err != nil {
+		return errors.Wrap(err, "configuring test log directory watcher to ignore the test log spec file")
+	}
+
+	started := make(chan struct{})
+	go func() {
+		h.watcher.Wait()
+		close(started)
+	}()
+	startErr := make(chan error)
+	go func() {
+		startErr <- h.watcher.Start(time.Millisecond)
```

Comment:
Yes, good catch! I hate the way this start mechanism returns an error but also blocks on success, instead of spawning a goroutine and returning immediately. Alas, this strange workaround.